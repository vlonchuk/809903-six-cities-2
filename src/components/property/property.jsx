import React, {PureComponent} from 'react';
import {connect} from "react-redux";
import PropTypes from 'prop-types';
import PageHeader from './../page-header/page-header.jsx';
import ActionCreator from './../../reducer/action-creator/action-creator.js';
import Operation from './../../reducer/operation/operation.js';
import {
  getPropertyById,
  convertRatingToPercent,
  getPropertiesByCity,
  getNearbyPlaces,
  showError,
} from './../../utils.js';
import {
  MAX_IMAGE_COUNT,
  MAX_NEARBY_PLACES,
} from '../../consts/constraints.js';
import ReviewList from './../review-list/review-list.jsx';
import Map from './../map/map.jsx';
import PlacesList from './../places-list/places-list.jsx';
import PlacesListType from './../../consts/places-list-type.js';
import Errors from './../../consts/errors.js';
import {Widths, Heights} from './../../consts/style.js';

class Property extends PureComponent {
  constructor(props) {
    super(props);
    this._handleAddToFavorite = this.handleAddToFavorite.bind(this);
  }

  handleAddToFavorite() {
    this.props.onAddToFavorite(this.props.property.id, this.props.property.isFavorite ? 0 : 1);
  }

  render() {
    return <div className="page">
      <PageHeader user={this.props.user}/>
      <main className="page__main page__main--property">
        <section className="property">
          {this.renderProperty()}
          <Map mapClass="property__map" properties={this.props.mapProperties}/>
          {this.renderNearbyProperties()}
        </section>
      </main>
    </div>;
  }

  renderNearbyProperties() {
    return <div className="container">
      <section className="near-places places">
        <h2 className="near-places__title">Other places in the neighbourhood</h2>
        <PlacesList listType={PlacesListType.NEAR} key="PlacesList" properties={this.props.properties} />
      </section>
    </div>;
  }

  renderProperty() {
    const {property} = this.props;
    const bookmarkClass = `property__bookmark-button button` + (property.isFavorite ? ` property__bookmark-button--active` : ``);
    const roundedRating = Math.round(property.rating);
    const ratingWidth = convertRatingToPercent(roundedRating);

    return <React.Fragment>
      <div className="property__gallery-container container">
        <div className="property__gallery">
          {
            property.images.slice(0, MAX_IMAGE_COUNT).map((el, i) =>
              <div className="property__image-wrapper" key={i}>
                <img className="property__image" src={el} alt="Photo studio"></img>
              </div>
            )
          }
        </div>
      </div>
      <div className="property__container container">
        <div className="property__wrapper">
          {
            property.isPremium ?
              <div className="property__mark">
                <span>Premium</span>
              </div>
              : null
          }
          <div className="property__name-wrapper">
            <h1 className="property__name">
              {property.title}
            </h1>
            <button className={bookmarkClass} type="button" onClick={this._handleAddToFavorite}>
              <svg className="property__bookmark-icon" width={Widths.PROPERTY_BOOKMARK_ICON} height={Heights.PROPERTY_BOOKMARK_ICON}>
                <use xlinkHref="#icon-bookmark"></use>
              </svg>
              <span className="visually-hidden">To bookmarks</span>
            </button>
          </div>

          <div className="property__rating rating">
            <div className="property__stars rating__stars">
              <span style={{width: ratingWidth + `%`}}></span>
              <span className="visually-hidden">Rating</span>
            </div>
            <span className="property__rating-value rating__value">{property.rating}</span>
          </div>

          <ul className="property__features">
            <li className="property__feature property__feature--entire">
              {property.type}
            </li>
            <li className="property__feature property__feature--bedrooms">
              {property.bedrooms} Bedroom{property.bedrooms > 1 ? `s` : ``}
            </li>
            <li className="property__feature property__feature--adults">
              Max {property.maxAdults} adult{property.maxAdults > 1 ? `s` : ``}
            </li>
          </ul>
          <div className="property__price">
            <b className="property__price-value">&euro;{property.price}</b>
            <span className="property__price-text">&nbsp;night</span>
          </div>

          <div className="property__inside">
            <h2 className="property__inside-title">What&apos;s inside</h2>
            <ul className="property__inside-list">
              {property.goods.map((el, i) =>
                <li className="property__inside-item" key={i}>
                  {el}
                </li>
              )}
            </ul>
          </div>
          {this.renderHost()}
          {this.renderComments()}
        </div>
      </div>
    </React.Fragment>;
  }

  renderHost() {
    const {property} = this.props;
    const avatarClass = `property__avatar-wrapper user__avatar-wrapper` + (property.host.isPro ? ` property__avatar-wrapper--pro` : ``);
    return <div className="property__host">
      <h2 className="property__host-title">Meet the host</h2>
      <div className="property__host-user user">
        <div className={avatarClass}>
          <img className="property__avatar user__avatar" src={`/` + property.host.avatarUrl} width={Widths.PROPERTY_AVATAR} height={Heights.PROPERTY_AVATAR} alt="Host avatar"></img>
        </div>
        <span className="property__user-name">
          {property.host.name}
        </span>
        {property.host.isPro ?
          <span className="property__user-status">
            Pro
          </span>
          : null
        }
      </div>
      <div className="property__description">
        <p className="property__text">
          {property.description}
        </p>
      </div>
    </div>;
  }

  renderComments() {
    return <ReviewList hotelId={this.props.id}/>;
  }
}

Property.propTypes = {
  id: PropTypes.number.isRequired,
  user: PropTypes.object,
  property: PropTypes.object,
  properties: PropTypes.array.isRequired,
  mapProperties: PropTypes.array.isRequired,
  onAddToFavorite: PropTypes.func,
};

const getNearbyProperties = (property, offers) => {
  const properties = getPropertiesByCity(property.city.name, offers);
  const propertiesExceptCurrent = properties.filter((el) => el.id !== property.id);
  return getNearbyPlaces(propertiesExceptCurrent, property, MAX_NEARBY_PLACES);
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  user: state.user,
});

const mapDispathToProps = (dispatch) => ({
  onAddToFavorite: (hotelId, status) => {
    dispatch(Operation.addToFavorite(hotelId, status));
  },
});

Property.getLinkProps = (offers, ownProps) => {
  const id = Number.parseInt(ownProps.match.params.id, 10);
  const property = getPropertyById(offers, id);

  if (!property) {
    return null;
  }

  const properties = getNearbyProperties(property, offers);
  const mapProperties = [property].concat(properties);

  return {
    id,
    property,
    properties,
    mapProperties,
  };
};

Property.loadParams = (dispatch, hotelId, property) => {
  dispatch(ActionCreator.activateCard(property));
  dispatch(Operation.loadComments(hotelId))
    .catch((err) => showError(err, Errors.ERR_LOAD_COMMENTS));
};

const PropertyWrapped = connect(mapStateToProps, mapDispathToProps)(Property);
export {Property};
export default PropertyWrapped;
