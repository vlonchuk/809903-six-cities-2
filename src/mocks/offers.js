const offers = [
  {
    id: `prop-1`,
    city: `Amsterdam`,
    caption: `Beautiful & luxurious apartment at great location`,
    imgSrc: `img/apartment-01.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 120,
    priceText: `night`,
    rating: 50,
    coor: {
      latitude: 52.3909553943508,
      longitude: 4.85309666406198
    },
  },
  {
    id: `prop-2`,
    city: `Amsterdam`,
    caption: `Wood and stone place`,
    imgSrc: `img/room.jpg`,
    type: `Private room`,
    priceCurrency: `€`,
    priceValue: 80,
    priceText: `night`,
    rating: 80,
    coor: {
      latitude: 52.369553943508,
      longitude: 4.85309666406198
    },
  },
  {
    id: `prop-3`,
    city: `Amsterdam`,
    caption: `Canal View Prinsengracht`,
    imgSrc: `img/apartment-02.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 132,
    priceText: `night`,
    rating: 100,
    coor: {
      latitude: 52.3909553943508,
      longitude: 4.929309666406198
    },
  },
  {
    id: `prop-4`,
    city: `Amsterdam`,
    caption: `Nice, cozy, warm big bed apartment`,
    imgSrc: `img/apartment-03.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 180,
    priceText: `night`,
    rating: 65,
    coor: {
      latitude: 52.3809553943508,
      longitude: 4.939309666406198
    },
  },
  {
    id: `prop-5`,
    city: `Paris`,
    caption: `Canal View Prinsengracht`,
    imgSrc: `img/apartment-02.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 132,
    priceText: `night`,
    rating: 90,
    coor: {
      latitude: 52.3909553943508,
      longitude: 4.929309666406198
    },
  },
  {
    id: `prop-6`,
    city: `Paris`,
    caption: `Nice, cozy, warm big bed apartment`,
    imgSrc: `img/apartment-03.jpg`,
    type: `Aparment`,
    priceCurrency: `€`,
    priceValue: 180,
    priceText: `night`,
    rating: 40,
    coor: {
      latitude: 52.3809553943508,
      longitude: 4.939309666406198
    },
  },
];

export const getPropertiesByCity = (city) => {
  return offers.filter((el) => el.city === city)
               .slice();
};

export default offers;
