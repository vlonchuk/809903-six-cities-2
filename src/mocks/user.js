const user = {
  id: 1,
  email: `Oliver.conner@gmail.com`,
  name: `Oliver.conner`,
  avatarUrl: `img/1.png`,
  isPro: false
};

export default user;
