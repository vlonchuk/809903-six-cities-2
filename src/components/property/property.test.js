import React from 'react';
import renderer from 'react-test-renderer';
import {Property} from './property';
import offers from './../../mocks/offers.js';
import user from './../../mocks/user.js';
import {BrowserRouter as Router} from 'react-router-dom';

jest.mock(`./../review-list/review-list.jsx`, () => jest.fn().mockReturnValue(null));
jest.mock(`./../map/map.jsx`, () => jest.fn().mockReturnValue(null));
jest.mock(`./../places-list/places-list.jsx`, () => jest.fn().mockReturnValue(null));

it(`ReviewAdd correctly renders after relaunch`, () => {
  const tree = renderer
    .create(
        <Router>
          <Property
            id={1}
            user={user}
            property={offers[0]}
            properties={offers}
            mapProperties={offers}
            onAddToFavorite={jest.fn()}
          />
        </Router>
    ).toJSON();
  expect(tree).toMatchSnapshot();
});
