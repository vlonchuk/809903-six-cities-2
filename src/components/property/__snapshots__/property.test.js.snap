// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewAdd correctly renders after relaunch 1`] = `
<div
  className="page"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height={41}
              src="/img/logo.svg"
              width={81}
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/favorites"
                onClick={[Function]}
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  className="header__user-name user__name"
                >
                  Oliver.conner@gmail.com
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        >
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="img/1.png"
            />
          </div>
          <div
            className="property__image-wrapper"
          >
            <img
              alt="Photo studio"
              className="property__image"
              src="img/2.png"
            />
          </div>
        </div>
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            >
              Beautiful & luxurious studio at great location
            </h1>
            <button
              className="property__bookmark-button button"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="property__bookmark-icon"
                height={33}
                width={31}
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "100%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
            <span
              className="property__rating-value rating__value"
            >
              4.8
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            >
              apartment
            </li>
            <li
              className="property__feature property__feature--bedrooms"
            >
              3
               Bedroom
              s
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              Max 
              4
               adult
              s
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              120
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <ul
              className="property__inside-list"
            >
              <li
                className="property__inside-item"
              >
                Heating
              </li>
              <li
                className="property__inside-item"
              >
                Kitchen
              </li>
              <li
                className="property__inside-item"
              >
                Cable TV
              </li>
              <li
                className="property__inside-item"
              >
                Washing machine
              </li>
              <li
                className="property__inside-item"
              >
                Coffee machine
              </li>
              <li
                className="property__inside-item"
              >
                Dishwasher
              </li>
            </ul>
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper user__avatar-wrapper property__avatar-wrapper--pro"
              >
                <img
                  alt="Host avatar"
                  className="property__avatar user__avatar"
                  height={74}
                  src="/img/1.png"
                  width={74}
                />
              </div>
              <span
                className="property__user-name"
              >
                Angelina
              </span>
              <span
                className="property__user-status"
              >
                Pro
              </span>
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              >
                A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        className="container"
      >
        <section
          className="near-places places"
        >
          <h2
            className="near-places__title"
          >
            Other places in the neighbourhood
          </h2>
        </section>
      </div>
    </section>
  </main>
</div>
`;
